### ISSUE TITLE
Large memory reallocation by Arrow causes hang in jemalloc

### ISSUE THREAD

**Issue by toddfarmer** [toddfarmer/arrow-migration#1252]
***Note**: This issue was originally created as [ARROW-1282](https://issues.apache.org/jira/browse/ARROW-1282). Please see the [migration documentation](https://gist.github.com/toddfarmer/12aa88361532d21902818a6044fda4c3) for further details.*

### Original Issue Description:
When reallocating a large amount of memory, Arrow is either triggering a bug in jemalloc or has a bug itself in the memory manager (many different applications reporting same issue but not clear from jemalloc issue description if they're sure it's in jemalloc or caused by other issues like using multiple memory allocation libraries in the same process, multithreaded access, etc).

Link to stack trace is here: https://gist.github.com/jeffknupp/73879feacf9c560afd4f1a20213dc6ef

Link to issue in jemalloc GitHub is here: https://github.com/jemalloc/jemalloc/issues/802

Originally observed in redis, discussed with jemalloc maintainer here: https://github.com/antirez/redis/issues/3799

**This is entirely reproducible on Ubuntu 16.04 xenial, which uses version 3.6.0 according to `apt` metadata.**

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16102671) by Wes McKinney (wesm):*
Is this reproducible in jemalloc 4.4.0? We are pinned at 4.4.0 in conda-forge:

https://github.com/conda-forge/arrow-cpp-feedstock/blob/master/recipe/meta.yaml#L38

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16102675) by Wes McKinney (wesm):*
`[~xhochy]` how are we handling jemalloc in manylinux1 builds?

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16102678) by Jeff Knupp (jeff.knupp@enigma.com):*
It’s reproducible in 4.5.0 I’ve yet to test against 5.0 which was only released about a month ago, but I’m assuming it is also present in 0.4.4. I’ll try to install conda on one of our instances and test; the only issue is I’m using the dev version of pandas (which has my fix for the crashes that were happening on large columns) and I don’t have a lot of experience using condas, but I’ll do my best.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16103156) by Uwe Korn (uwe):*
In the manylinux1 builds, we pin to `jemalloc==4.4.0`.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16103190) by Wes McKinney (wesm):*
Statically linking to 4.4.0, right?

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16103281) by Wes McKinney (wesm):*
`[~jeffknupp]` you're definitely hitting this in jemalloc 4.5.0, i.e. including this fix https://github.com/jemalloc/jemalloc/commit/adae7cfc4a2ac66c96b0dcc83b3837ac668fc44e ? 

If there's a way we can reproduce this we can do some debugging on Ubuntu 16.04. There are some other issues preventing us from upgrading to 5.0.x at the moment (https://github.com/jemalloc/jemalloc/issues/937)

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16103337) by Jeff Knupp (jeffknupp):*
I'm hitting this with both 4.5 **and** 5.0.1. There was talk of this being related to the application's use of jemalloc along with another allocator and/or in a multithreaded environment causing corruption of the jemalloc datastructures, so it's possible it's not a bug that jemalloc has/will fix. I have to do more debugging, but I'll also try to set up a Dockerfile to reproduce this (the biggest issue is that it requires reading and converting a number of large CSVs for a few minutes before it happens, so I need to provide some large fake data files and make sure they still trigger the issue).

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16103382) by Jeff Knupp (jeffknupp):*
Correction, not 5.0.1 as it's not compatible (as you mentioned), but definitely with 4.5.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16103395) by Wes McKinney (wesm):*
I see. From the comment at https://github.com/antirez/redis/issues/3799#issuecomment-297832432 this is definitely concerning since there are plenty of other libraries in the stack that will not be using jemalloc. Particularly with multithreaded Parquet reads, there might be two threads which are using different allocators (e.g. the Python interpreter is using the system allocator, while Parquet/Arrow are using jemalloc)

`[~xhochy]` the conservative approach on this might be to disable jemalloc in our conda and PyPI packages until we understand what is going on. What do you think?

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16103766) by Jeff Knupp (jeff.knupp@enigma.com):*
Probably safest to roll back. I think I'm pretty close to determining the
cause but probably better not to have a broken version out.

On Thu, Jul 27, 2017 at 12:09 PM, Wes McKinney (JIRA) <jira@apache.org>

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16103796) by Wes McKinney (wesm):*
Just opened https://github.com/conda-forge/arrow-cpp-feedstock/pull/35. Will need to build new wheels for PyPI. Luckily `[~xhochy]` made it really easy to turn the jemalloc allocator on and off

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16103806) by Jeff Knupp (jeffknupp):*
FWIW, it definitely appears to be an issue caused by other allocators calling `sbrk()`, which it seems `jemalloc` doesn't play nicely with. I'm wondering if it would be fixed by recompiling my Python interpreter to use jemalloc as well. Regardless, I'm still investigating and will try to have a way for everyone to reproduce in the next day or two.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16103824) by Wes McKinney (wesm):*
I appreciate it. I'm a bit spooked by this issue, in the jemalloc wiki it suggests that having applications with both jemalloc and system allocator is OK

https://github.com/jemalloc/jemalloc/wiki/Getting-Started

I wonder if this problem goes away / gets better if we compile jemalloc with a prefix and use the prefixed functions in our memory allocator. Redis does this https://github.com/antirez/redis/blob/unstable/deps/Makefile#L80

As I understand it, if you compile jemalloc unprefixed, then the intended use is to override all malloc/free calls in the application (e.g. using LD_PRELOAD) so that libc malloc/free never get used

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16104256) by Jeff Knupp (jeffknupp):*
`[~wesmckinn]` `[~xhochy]` Actually, we may want to hold off on removing jemalloc for a moment. In testing `Arrow::from_pandas` on a large dataframe (4GB CSV file), the allocation strategy used without `jemalloc` is _very_ conservative and only resizes to fit the next element in a call to `BufferBuilder::append`. This essentially slows the program down so much as to be unusable. We should probably come up with a better way of signaling from `ConvertObjectStrings`/`AppendObjectStrings` a rough estimate of the total size we'll need (which we should know since we already have the complete dataframe in memory, right?).

Sorry, I'm trying everything I can to get this all to work together, but it seems that my usage (very large Tables) is non-typical, so I'm hitting every edge-case there is.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16104263) by Wes McKinney (wesm):*
Oh, that's a bug that I introduced recently, and would have been caught if we were benchmarking more rigorously. We should be doubling the buffer size when we run out of capacity in BufferBuilder instead of only expanding to fit the next-appended bit. I think this flew under the radar when I changed `BinaryBuilder` from using a `UInt8Builder` under the hood (which does array doubling to grow the arrays). 

I opened https://issues.apache.org/jira/browse/ARROW-1290. I am betting that this will reduce the number of calls to `buffer->Resize` in BufferBuilder and maybe make this problem go away for you

Seems we may want to do Arrow 0.6.0 very soon including this fix as soon we can get the Plasma code ready to ship.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16104450) by Wes McKinney (wesm):*
As another thing to try, what happens when you compile jemalloc with a prefix and statically link? So we would add something like `je_arrow_` to all of the jemalloc functions in https://github.com/apache/arrow/blob/master/cpp/src/arrow/memory_pool.cc. Because when libjemalloc is loaded it may interfere with the libc malloc and free functions, this might make them play more nicely together. 

So you would set ARROW_JEMALLOC_USE_SHARED=off in https://github.com/apache/arrow/blob/master/cpp/CMakeLists.txt#L105 and then add the prefix in the thirdparty build like in Redis https://github.com/antirez/redis/blob/unstable/deps/Makefile#L80

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16105051) by Jeff Knupp (jeffknupp):*
Yeah, that's what I'm working on now. Just about finished making the c++ changes. Will report results soon.

Also, my process ran successfully for 12 hours without jemalloc (used to hang after 10 minutes), so at least the other fixes I had been waiting on are working.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16105639) by Jeff Knupp (jeffknupp):*
`[~wesmckinn]`, I tried using the jemalloc with a prefix and statically linked, but it the result was the same. I made sure by running in gdb that the `je_arrow_` prefixed symbols were present and that it was statically linked, so I think this is a dead-end.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16107562) by Jeff Knupp (jeffknupp):*
`[~wesmckinn]`, despite it not fixing the issue, it seems that using a prefix it still clearly the correct way to use jemalloc along with another allocator. Do you want me to submit a PR for the changes I made?

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16107596) by Wes McKinney (wesm):*
Yes, please. You can create a new JIRA for this change. Maybe `je_arrow_` to avoid conflicts. We may have to do some work in our build toolchain because we have various places where we use the system jemalloc. `[~xhochy]` is more of an expert in this so I defer to his guidance

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16111371) by Chris Bartak (chrisb1):*
I've run into a problem that I assume has to be this issue, unfortunately I can't quite get it down to a reproducible example, but I'll share the context in case it's helpful.

ec2 2017.03 Amazon Linux AMI  (Red Hat 4.8.3-9)
python3.4
pyarrow==0.5  (from pip + deps)

Reading a very small parquet file, 3kb - two text columns.  Interactively seems to always work.  Serving a webapp with `httpd`/`mod_wsgi` and Flask that reads the same file - almost always (but not always!) it completely hangs.  No spike in CPU/memory

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16111541) by Wes McKinney (wesm):*
OK. I'm going to get to work putting up patched 0.5.0 builds on PyPI and conda-forge since these issues persisting is not acceptable. We should still figure out what is happening in jemalloc to cause this but it may take a little while

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16111614) by Wes McKinney (wesm):*
Moving this issue to 0.7.0 as it doesn't seem likely the underlying cause will be resolved in time for 0.6.0. I created ARROW-1312 to switch off the allocator by default to triage the situation.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16111793) by Wes McKinney (wesm):*
`[~chrisb1]` I have just posted a patched build on PyPI (0.5.0.post1, because PyPI does not allow new builds with the same version number). If you install with

`pip install pyarrow==0.5.*`

then this issue should go away. Please let me know if not

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16111798) by Wes McKinney (wesm):*
I made a mistake in the build settings, will post a new set of binaries within a half hour or so.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16111903) by Chris Bartak (chrisb1):*
I'm still seeing my issue with `pyarrow==0.5.0.post2`, must be something else.  Certainly possible that it's unrelated to pyarrow, though I thought I had it pretty well isolated.  I'll open  a new issue if I can get it reproducible.  Thanks for the quick upload!

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16111935) by Wes McKinney (wesm):*
Great, thank you. If you can provide a gdb backtrace from the hung process that would be helpful

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16112693) by Chris Bartak (chrisb1):*
Here's a backtrace - I guess it's actually hanging on re-acquiring the GIL - any theories?  

```Java
Thread 2 (Thread 0x7fe5e97ab700 (LWP 12562)):
#0  0x00007fe5f6dcda82 in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007fe5eb0f1953 in ?? () from /usr/lib64/libpython3.4m.so.1.0
#2  0x00007fe5eb0f1f7b in PyEval_RestoreThread () from /usr/lib64/libpython3.4m.so.1.0
#3  0x00007fe5eb11729b in PyGILState_Ensure () from /usr/lib64/libpython3.4m.so.1.0
#4  0x00007fe5d0fa143d in __pyx_f_7pyarrow_3lib_check_status(arrow::Status const&) () from /opt/python/run/venv/lib64/python3.4/site-packages/pyarrow/lib.cpython-34m.so
#5  0x00007fe5d101c040 in __pyx_pw_7pyarrow_3lib_16MemoryMappedFile_5_open(_object*, _object*, _object*) ()
   from /opt/python/run/venv/lib64/python3.4/site-packages/pyarrow/lib.cpython-34m.so
#6  0x00007fe5d0fc4198 in __pyx_pw_7pyarrow_3lib_73memory_map(_object*, _object*, _object*) ()
   from /opt/python/run/venv/lib64/python3.4/site-packages/pyarrow/lib.cpython-34m.so
#7  0x00007fe5d0fb673d in __pyx_f_7pyarrow_3lib_get_reader(_object*, std::shared_ptr<arrow::io::RandomAccessFile>*) ()
   from /opt/python/run/venv/lib64/python3.4/site-packages/pyarrow/lib.cpython-34m.so
#8  0x00007fe5d02303ce in __pyx_pw_7pyarrow_8_parquet_13ParquetReader_3open(_object*, _object*, _object*) ()
   from /opt/python/run/venv/lib64/python3.4/site-packages/pyarrow/_parquet.cpython-34m.so
#9  0x00007fe5eb0fa765 in PyEval_EvalFrameEx () from /usr/lib64/libpython3.4m.so.1.0
#10 0x00007fe5eb0fbdee in PyEval_EvalCodeEx () from /usr/lib64/libpython3.4m.so.1.0
#11 0x00007fe5eb06e1b3 in ?? () from /usr/lib64/libpython3.4m.so.1.0
```

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16112706) by Wes McKinney (wesm):*
Awesome, yeah! The GIL is not being released here https://github.com/apache/arrow/blob/master/python/pyarrow/io.pxi#L450 which is apparently causing a deadlock when run in a background thread. I wasn't aware this could happen

Opened https://issues.apache.org/jira/browse/ARROW-1327 to track this fix (I would also like to see if coding errors like this can be made preventable)

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16114616) by Uwe Korn (uwe):*
Notes while digging on this issue:

- I can reproduce this in the environment in https://github.com/xhochy/arrow-dockerfiles/tree/master/ARROW-1282 using the NYC Taxi Tripdata CSV from January 2016 using only VendorID and all columns that would be `object` on using plain `pd.read_csv`
- Using the script `hanger.py` in there, it always hangs on the third iteration.
- Building `jemalloc==4.5.0` using `--with-malloc-conf=dss:disabled` avoids the issue.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16114734) by Jeff Knupp (jeff.knupp@enigma.com):*
Fantastic! I wasn't aware of the `dss:disabled` flag but it makes sense
that that would fix it as that's where the issue was. Is there anything
stopping us from moving to 4.5 and statically linking with that compile
option?

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16114743) by Wes McKinney (wesm):*
I think the idea is that this disables `sbrk`, but this function was only ever being called when a `mmap`-based memory allocation failed. I would prefer to put a stable 0.6.0 release out and then experiment with these changes for a couple of weeks before making it the default. We can go ahead and change the thirdparty toolchain in Arrow to use jemalloc 4.5.0 and build with this config flag, since the default is not to use it. `[~jeffknupp]` since you have a patch going already do you want to include this change in it?

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16114752) by Jeff Knupp (jeffknupp):*
Sure, will do.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16143080) by Uwe Korn (uwe):*
Current state: It does not seem like large reallocations are a problem but rather certain sequences. I have now a reproducible environment with `jemalloc==4.5.0` where I can reproduce the hangs. The stacktrace is similar to that of `[~jeffknupp]` for the jemalloc part but different for the Arrow trace. So this points to a jemalloc issue and not in the Arrow codebase.

Stacktrace: https://gist.github.com/xhochy/b3d484807ab68814092d648adb6e1ff5

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16143176) by Uwe Korn (uwe):*
There is also now a docker environment to reproduce the bugs consistently: https://github.com/xhochy/arrow-dockerfiles/tree/master/ARROW-1282 I will start playing there with the jemalloc source to better understand the problem.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16143201) by Uwe Korn (uwe):*
The `mmap` calls are failing internally in jemalloc as `mmap` returns memory but not at the required memory address. These already happens several times before the hanger, so this seems unrelated.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16143803) by Jeff Knupp (jeffknupp):*
Wouldn't this be caused by `dss:disabled` as it affects `mmap` allocations?

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16143805) by Uwe Korn (uwe):*
`[~jeffknupp]` no, I'm running with `dss:enabled` at the moment. I would like to have this fixed for the default jemalloc configuration.

By default, jemalloc first tries to allocate a new page using `mmap`, if this fails to produce the additional memory at the specified location, `jemalloc` will try to allocate the memory using `sbrk`.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16144963) by Uwe Korn (uwe):*
Made a PR that fixes this for the jemalloc 4.x series: https://github.com/jemalloc/jemalloc/pull/1005

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16156365) by Wes McKinney (wesm):*
`[~xhochy]` what do you want to do about this issue? We would have to do an ExternalProject build at a pinned commit and statically link. Perhaps we can delay this until 0.8.0

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16156732) by Uwe Korn (uwe):*
Delayed until 0.8.0, I hope to see the PR merged soon upstream and then we could do a bit more testing before we enable it again.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1282?focusedCommentId=16243046) by Uwe Korn (uwe):*
Merged upstream and the thirdparty builds all pull the new revision. Will deal with the vendoring in a separate story.
